### PR TITLE
Override the decoder score method

### LINKED ISSUE #2667: Decoder.score() behaviour in classification mode
Hello!

I'm not sure if this issue is a documentation or feature request issue: the `.score` seems a bit confusing for me as a new-comer. 

I used the Decoder object in a classification tasks. After fitting the model, I used the  `.score` method, expecting it to compute the mean accuracy (as it is done in scikit-learn for classifiers). I got lucky its value came back negative, it made me look at the documentation and the fact that it computes R^2 coefficient.

I think it should be a bit more explicit to the user that `.score` in a classification fashion does not compute the mean accuracy:

- either by changing the behaviour (which seems tricky as it comes from inherited scikit-learn RegressorMixin)
- or by explicitly warn the user when running the method (which should be easier by using the `self.is_classification` flag) and make a visible warning in the documentation

### PR DESCRIPTION
**Pull request by NicolasGensollen**
Fixes #2667 

This PR proposes to override the `score` method of the base decoder class such as to compute scores using the scoring strategy provided by the user when instantiating the decoder rather than a default method. 

The scoring strategy of the decoder is defined by

- `decoder.scoring` : Callable, or str, or `None`. Either the scoring method itself or its name as defined in the docstring. If set to `None`, `decoder.score(X, y)` will raise an error.
- `decoding._scoring_metric` : Callable. The scoring method.

### CODE REVIEW

File: nilearn/decoding/decoder.py [hunk 1]
```diff
@@ -583,6 +617,33 @@ def fit(self, X, y, groups=None):
...
+        Returns
+        -------
+        score : float
+            Prediction score.
+
+        """
+        if self.scoring is None:
+            raise ValueError("Unable to compute score. "
+                             "No scoring metric was provided "
+                             "to Decoder object.")
```
Comment (bthirion) [inline nilearn/decoding/decoder.py:644]: But in that case, why not relay on a default metric depending on the nature of the classifier ? e.g. accuracy for a classifier, r2 for a regressor.

Comment (bthirion) [reply nilearn/decoding/decoder.py:644]: I feel I'm missing something, sorry...

Comment (RaphaelMeudec) [reply nilearn/decoding/decoder.py:644]: Suggestions (that might be naive as I'm not used to nilearn/sklearn) regarding the way to get the scorer:

- maybe we could use the `get_scorer` method instead of storing a list of available scorers on nilearn side?
- then we could have a `default_scorer` with value `"accuracy"` or `"r2"` depending on the `self.is_classification` flag

Comment (NicolasGensollen) [reply nilearn/decoding/decoder.py]: Thanks for the reviews @bthirion @RaphaelMeudec!

You're right, using `get_scorer` is much better :+1:. For some reason I thought that the string values of `decoder.scoring` wouldn't be valid sklearn scorer keys, but they totally are... 

>I feel I'm missing something, sorry...

I don't think so. I hesitated between relying on a default scoring metric and raising an error, and went for the last option to make sure the user doesn't get unexpected results. I agree this is over-restrictive and changed the behaviour to rely on a default metric depending on the nature of the classifier like you both suggested.

**Review by bthirion — COMMENTED** [review]
LGTM. beware that one line is not tested apparently.
Also, this deserves a whatsnew entry.

File: nilearn/decoding/decoder.py [hunk 2]
```diff
@@ -387,6 +388,15 @@ def __init__(self, estimator='svc', mask=None, cv=10, param_grid=None,
         self.n_jobs = n_jobs
         self.verbose = verbose
 
+        # Determine the scoring strategy
+        if self.scoring is not None:
+            self._scorer = get_scorer(self.scoring)
+        else:
```
Comment (jeromedockes) [inline nilearn/decoding/decoder.py:394]: should `fit` below use `self._scorer` instead of calling `check_scoring`? see:
```
 scorer = check_scoring(self.estimator, self.scoring)
```

although in any case it will not happen often that scoring is None, as users are not supposed to instantiate that class directly, but use Decoder or DecoderRegressor, that set defaults for `scoring`

Comment (NicolasGensollen) [reply nilearn/decoding/decoder.py]: Good point, thanks! 
I updated the code to setup the scorer in `fit` rather than in the constructor, and made `scorer` an attribute of the decoder.

Comment (jeromedockes) [reply nilearn/decoding/decoder.py]: looks great, thanks! one last detail is that I believe in scikit-learn's convention any attribute created during `fit` has a name ending with `_`, and the `__init__` does nothing besides storing the parameters passed to it. (I think it should be like this for `sklearn.base.clone` to work properly). So maybe it would stick more closely to sklearn's conventions if we did not initialize `scorer = None` in `__init__`, and added a `scorer_` attribute in `fit`?

**Review by jeromedockes — APPROVED** [review]
thanks!

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Comment by NicolasGensollen** [comment]
Thanks for the reviews @bthirion @jeromedockes @RaphaelMeudec ! :+1: 
Merging...
